File: python/src/problems1_100/valid_parentheses.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def is_valid(self, s: str) -> bool:
        pass


class Default(Problem):
    @override
    def is_valid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False

        symbols: list[str] = []
        for c in s:
            if c not in {')', '}', ']'}:
                symbols.append(c)
                continue

            if len(symbols) > 0 and \
                ((symbols[-1] == '(' and c == ')') or
                 (symbols[-1] == '{' and c == '}') or
                 (symbols[-1] == '[' and c == ']')):
                symbols.pop()
                continue
            return False
        return len(symbols) == 0


SOLUTIONS: list[Problem] = [Default()]

File: javascript/src/problems1_100/valid-parentheses.js
class Problem {
  isValid(s) {
    throw new Error(`Not implemented (${s}).`);
  }
}

class Default extends Problem {
  isValid(s) {
    if (s.length % 2 === 1) {
      return false;
    }

    const symbols = [];
    for (const c of s) {
      if (!(c === ')' || c === '}' || c === ']')) {
        symbols.push(c);
        continue;
      }

      if (symbols.length > 0 &&
          ((symbols[symbols.length - 1] === '(' && c === ')') ||
              (symbols[symbols.length - 1] === '{' && c === '}') ||
              (symbols[symbols.length - 1] === '[' && c === ']'))
      ) {
        symbols.pop();
        continue;
      }
      return false;
    }
    return symbols.length === 0;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
